### PR TITLE
Feat: MCP Traffic gateway

### PR DESCRIPTION
**Pull request by tomek-labuk**
## Description

Addresses [issue 1365](https://github.com/Kong/developer.konghq.com/issues/1365#issue-3038918758). This PR adds a new section about securing, governing and observing MCP traffic.

**NOTES for reviewers:**
* For testing, I used my own repository. I can share the repository link and a personal access token (PAT) via 1Password. One open question: should we provide a dummy repository for users to fork, or assume they’ll create or use their own test repositories?
* The how-to guides use the new `route_type: llm/v1/responses`, which leverages the [[OpenAI Responses API](https://cookbook.openai.com/examples/mcp/mcp_tool_guide)](https://cookbook.openai.com/examples/mcp/mcp_tool_guide). This allows remote MCP servers to be used without requiring users to host them. Currently, Kong AI Gateway and Kong Gateway don't support proxying MCP servers directly.
* The how-to guides are designed as a related series. One missing piece in the observability guide is a section on visualizing metrics in Grafana. I’m waiting on LLM metrics to be added to the official Kong Dashboard (Antoine will handle this after his PTO).
* The Konnect MCP Server is included to clarify that it’s officially supported as part of Kong’s MCP/AI Gateway offering.
* You will notice that the responses from the /responses API are long, as they return a list of available tools and their usage. Look for `"type": "message"` in the JSON response.

Let me know if you'd like a more formal or internal-facing tone.

## Preview Links

https://deploy-preview-1925--kongdeveloper.netlify.app/mcp/
https://deploy-preview-1925--kongdeveloper.netlify.app/mcp/secure-mcp-traffic/
https://deploy-preview-1925--kongdeveloper.netlify.app/mcp/govern-mcp-traffic/
https://deploy-preview-1925--kongdeveloper.netlify.app/mcp/observe-mcp-traffic/

https://deploy-preview-1925--kongdeveloper.netlify.app/mcp/get-started/
https://deploy-preview-1925--kongdeveloper.netlify.app/mcp/tools/

## Checklist 

- [x] Every page is page one.
- [x] Tested how-to docs. If not, note why here. 
- [x] All pages contain metadata.
- [ ] Updated [sources.yaml](https://github.com/Kong/developer.konghq.com/blob/main/tools/track-docs-changes/config/sources.yml). For more info, review [track docs changes](https://github.com/Kong/developer.konghq.com/tree/main/tools/track-docs-changes)
- [x] Any new docs link to existing docs.
- [x] All autogenerated instructions render correctly (API, decK, Konnect, Kong Manager). 
- [x] Style guide (capitalized gateway entities, placeholder URLs) implemented correctly.
- [x] Every page has a `description` entry in frontmatter
- [ ] Add new pages to the product documentation index (if applicable)

### CODE REVIEW

File: app/_how-tos/govern-mcp-traffic.md [hunk 1]
```diff
@@ -0,0 +1,457 @@
...
+
+{% endnavtabs %}
+
+## Enforce local rate limiting
+
+Sure! Here’s the updated intro with that note included:
+
+---
+
+This configuration enables the `ai-rate-limiting-advanced` plugin to enforce strict rate limits on requests to the OpenAI provider. It uses a **local strategy** with a **fixed window** of 10 seconds, allowing only **1 request per 10-second window**. Requests exceeding this limit within the same window will receive a `429 Too Many Requests` response, effectively controlling request bursts and protecting backend resources.
```
Comment (cloudjumpercat) [inline app/_how-tos/govern-mcp-traffic.md:378]:
```suggestion
Now you can enforce local rate limiting by configuring the AI Rate Limiting Advanced plugin to enforce strict rate limits on requests to the OpenAI provider. This uses a local strategy with a fixed window of 10 seconds, which only allows one request per 10-second window. Requests exceeding this limit within the same window will receive a `429 Too Many Requests` response, effectively controlling request bursts and protecting backend resources.
```

File: app/_how-tos/govern-mcp-traffic.md [hunk 2]
```diff
@@ -0,0 +1,457 @@
...
+
+{:.info}
+> **Note:** This configuration is for showcase and testing purposes only. In a production environment, rate limits and window sizes should be adjusted to match actual usage patterns and performance requirements.
```
Comment (cloudjumpercat) [inline app/_how-tos/govern-mcp-traffic.md:381]:
```suggestion
> **Note:** This configuration is for testing purposes only. In a production environment, rate limits and window sizes should be adjusted to match actual usage patterns and performance requirements.
```

File: app/_how-tos/govern-mcp-traffic.md [hunk 3]
```diff
@@ -0,0 +1,457 @@
...
+      config:
+        strategy: local
+        window_type: fixed
+        llm_providers:
+        - name: openai
+        window_size:
+        - 10
+        limit:
+        - 1
+{% endentity_examples %}
```
Comment (cloudjumpercat) [inline app/_how-tos/govern-mcp-traffic.md:397]: This gave me the following error:

```
Error: 1 errors occurred:
	while processing event: Create plugin ai-rate-limiting-advanced (global) failed: HTTP status 400 (message: "validation error: unknown field")
```

File: app/_how-tos/govern-mcp-traffic.md [hunk 4]
```diff
@@ -0,0 +1,457 @@
...
+
+
+## Validate rate limiting
+
+We can test our simple rate limiting configuration by running the following script:
+
+```bash
+for i in {1..6}; do
```
Comment (cloudjumpercat) [inline app/_how-tos/govern-mcp-traffic.md:405]: I'd use `{% validation rate-limit-check %}` for this one as well. https://developer.konghq.com/contributing/#rate-limit-check

File: app/_how-tos/govern-mcp-traffic.md [hunk 5]
```diff
@@ -0,0 +1,457 @@
...
+    }")
+  echo "$http_status"
+
+  if [[ $i == 3 ]]; then
+    sleep 10
+  else
+    sleep 1
+  fi
+done
+```
```
Comment (cloudjumpercat) [inline app/_how-tos/govern-mcp-traffic.md:433]: I got all 400s when I ran this.

File: app/_how-tos/observe-mcp-traffic.md [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+        ```bash
+        export GITHUB_PAT=<YOUR_GITHUB_TOKEN>
+        ```
+      icon_url: /assets/icons/third-party/github.svg
+  prereqs:
+    entities:
+        services:
+            - example-clean-service
+        routes:
+            - example-clean-route
```
Comment (cloudjumpercat) [inline app/_how-tos/observe-mcp-traffic.md:72]:
```suggestion
```
I think these can be removed since we created them earlier in the series

File: app/_how-tos/observe-mcp-traffic.md [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+      include_content: cleanup/platform/konnect
+      icon_url: /assets/icons/gateway.svg
+    - title: Destroy the {{site.base_gateway}} container
+      include_content: cleanup/products/gateway
+      icon_url: /assets/icons/gateway.svg
+---
+
+## Reconfigure the AI Proxy Advanced plugin
+
+Now we enable detailed logging for all configured models in the AI Proxy Advanced plugin. This captures request/response payloads and performance statistics. We then scrape those statistics using the Prometheus plugin for monitoring and analysis. Apply the configuration below to enable logging for both used models.
```
Comment (cloudjumpercat) [inline app/_how-tos/observe-mcp-traffic.md:86]:
```suggestion
Now, you can enable detailed logging for all configured models in the AI Proxy Advanced plugin. This captures request/response payloads and performance statistics. You can then scrape those statistics using the Prometheus plugin for monitoring and analysis. Apply the configuration below to enable logging for both models.
```

File: app/_how-tos/observe-mcp-traffic.md [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+              log_payloads: true
+              log_statistics: true
+            auth:
+              header_name: Authorization
+              header_value: Bearer ${openai_api_key}
+            weight: 50
+variables:
+  openai_api_key:
+    value: $OPENAI_API_KEY
+{% endentity_examples %}
```
Comment (cloudjumpercat) [inline app/_how-tos/observe-mcp-traffic.md:127]: This gave me the following error (RC7 on Konnect):

```
Error: 1 errors occurred:
	while processing event: Update plugin ai-proxy-advanced (global) failed: HTTP status 400 (message: "validation error: expected one of: llm/v1/chat, llm/v1/completions, preserve")
```

**Review by cloudjumpercat — CHANGES_REQUESTED** [review]
Left suggestions and comments, main things:
* Some extra commits/changes were accidentally brought in when I think you changed the base branch, make sure those are fixed
* I couldn't get the how tos to work for me, I've left the errors I got as comments. It wasn't clear when I was supposed to use the GitHub repo you had. As for if we should share that demo repo, I think that would depend on how it's used. I'm not sure how it's used at the moment.
* If you use gen AI, I find it's helpful to call it out with a comment for reviewers. I'll call out when I use gen AI for a curl example and ask someone, like Fabian or a dev, to verify that it is correct as well as test it myself.

File: app/_how-tos/get-started-with-mcp.md [hunk 9]
```diff
@@ -0,0 +1,370 @@
...
+# Optional: The API region to use (defaults to US)
+# Possible values: US, EU, AU, ME, IN
+export KONNECT_REGION=us
+```
+
+## Configure Claude Desktop
+
+Claude uses a configuration file to register custom MCP servers. You’ll need to create or edit this file based on your operating system:
+
+* **macOS**: `~/Library/Application Support/Claude/claude_desktop_config.json`
```
[the remark this answers is not in the document]
Comment (tomek-labuk) [reply app/_how-tos/get-started-with-mcp.md:150]: If that file isn't there, you just have to create it

File: app/_how-tos/observe-mcp-traffic.md [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+          \"headers\": {
+            \"Authorization\": \"Bearer $GITHUB_PAT\"
+          }
+        }
+      ],
+      \"input\": \"tools available with github mcp\"
+    }" | jq -r '.model'
+  sleep 3
+done
+```
```
[the remark this answers is not in the document]
Comment (tomek-labuk) [reply app/_how-tos/observe-mcp-traffic.md:209]: But your AI Proxy doesn't hit the /responses route_type, right?

File: app/_how-tos/secure-mcp-traffic.md [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+      - type: mcp
+        server_label: gitmcp
+        server_url: https://api.githubcopilot.com/mcp/
+        require_approval: never
+        headers:
+            Authorization: Bearer $GITHUB_PAT
+    input: how do i use github mcp
+  status_code: 200
+  message: 'OK'
+{% endvalidation %}
```
[the remark this answers is not in the document]
Comment (tomek-labuk) [reply app/_how-tos/secure-mcp-traffic.md:167]: I've just tested them, and I get 200s. I think it's the AI Proxy configuration that causes 400s, let me think. You don't need to pull that repo, you just need to mention it explicitly in your "input", like in the example: `"Create an issue in the repository YOUR_REPOSITORY_NAME with title: Example Issue Title and body: This is the description of the issue created via MCP."`

File: app/mcp/tools.md [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+tags:
+  - ai
+  - streaming
+
+plugins:
+  - ai-proxy
+  - ai-proxy-advanced
+
+min_version:
+  gateway: '3.6'
```
[the remark this answers is not in the document]
Comment (tomek-labuk) [reply app/mcp/tools.md:27]: It must be 3.11, let me change that. I put 3.10, as my permalinks were getting broken with 3.11 defined here

File: app/mcp/tools.md [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+related_resources:
+    - text: Kong MCP Konnect
+      url: 'https://github.com/Kong/mcp-konnect'
+    - text: Kong MCP Konnect on Docker
+      url: 'https://hub.docker.com/r/mcp/kong'
+      icon: /assets/icons/third-party/docker.svg
+    - text: AI Gateway
+      url: /ai-gateway/
+---
+
```
[the remark this answers is not in the document]
Comment (tomek-labuk) [reply app/mcp/tools.md:38]: Agreed, and I'm not even sure if that page is required. You can check these tools from Claude / GitHub repository.

File: app/_how-tos/secure-mcp-traffic.md [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+
+        ```bash
+        export GITHUB_PAT=<YOUR_GITHUB_TOKEN>
+        ```
+      icon_url: /assets/icons/third-party/github.svg
+  entities:
+    services:
+        - example-clean-service
+    routes:
+        - example-clean-route
```
[the remark this answers is not in the document]
Comment (tomek-labuk) [reply app/_how-tos/secure-mcp-traffic.md:74]: The reason here is that example-service has the following URL: url: http://httpbin.konghq.com/anything and example-route uses /anything path. This configuration will not work with the llm/v1/responses endpoint. It will only work when the url is set to http://httpbin.konghq.com/. I know that the team will work on fixing that behavior.

File: app/mcp/tools.md [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+Kong’s MCP Server provides a set of API-accessible tools to support analytics, configuration management, and control plane administration. These tools help users monitor traffic, query resource metadata, and manage control plane hierarchies programmatically.
```
[the remark this answers is not in the document]
Comment (tomek-labuk) [reply app/mcp/tools.md]: Thanks, that's fixed now

File: app/_how-tos/get-started-with-mcp.md [hunk 16]
```diff
@@ -0,0 +1,370 @@
...
+### Step 3: Query API Traffic for a Service
+
+To analyze traffic and detect error trends, run a query like this:
+
+{% navtabs "query-api-traffic" %}
+{% navtab "Prompt" %}
+
+```text
+Prompt:
+Show me all API requests for service [SERVICE_NAME or ID] in the last hour that had 5xx status codes.
```
[the remark this answers is not in the document]
Comment (tomek-labuk) [reply app/_how-tos/get-started-with-mcp.md]: I'll add it, good catch

File: app/_how-tos/govern-mcp-traffic.md (hunk 5 above)
Comment (tomek-labuk) [reply app/_how-tos/govern-mcp-traffic.md:409]: I tweaked this configuration and it should work better

File: app/_how-tos/observe-mcp-traffic.md [hunk 17]
```diff
@@ -0,0 +1,223 @@
...
+{: data-deployment-topology="on-prem" }
+
+```yaml
+scrape_configs:
+ - job_name: 'kong'
+   scrape_interval: 5s
+   static_configs:
+     - targets: ['kong-quickstart-gateway:8100']
+```
+{: data-deployment-topology="konnect" }
```
[the remark this answers is not in the document]
Comment (tomek-labuk) [reply app/_how-tos/observe-mcp-traffic.md:146]: I think that I'll skip Konnect in this iteration and I'll work on Konnect-related MCP features in July. There's a growing pressure to publish these docs.

File: app/_how-tos/get-started-with-mcp.md [hunk 18]
```diff
@@ -0,0 +1,372 @@
...
+
+### Query API traffic for a Service
+
+To analyze traffic and detect error trends, run a query like this:
+
+{% navtabs "query-api-traffic" %}
+{% navtab "Prompt" %}
+
+```text
+Show me all API requests for the example-service Gateway Service in the last hour that had 5xx status codes.
```
Comment (cloudjumpercat) [inline app/_how-tos/get-started-with-mcp.md:294]:
```suggestion
Show me all API requests for the example-service Gateway Service in the quickstart Control Plane in the last hour that had 5xx status codes.
```

File: app/_how-tos/get-started-with-mcp.md [hunk 19]
```diff
@@ -0,0 +1,372 @@
...
+{% endnavtabs %}
+
+## Troubleshoot Consumer issues
+
+### List Consumers in a Control Plane
+
+Start by getting the list of consumers for a control plane:
+
+```text
+List all Consumers for the quickstart control plane.
```
Comment (cloudjumpercat) [inline app/_how-tos/get-started-with-mcp.md:351]:
```suggestion
List all Consumers for the quickstart Control Plane.
```

File: app/_how-tos/get-started-with-mcp.md [hunk 20]
```diff
@@ -0,0 +1,372 @@
...
+        "KONNECT_REGION": "us"
+      }
+    }
+  }
+}
+```
+
+{:.warning}
+> Replace `/absolute/path/to/mcp-konnect/build/index.js` with the full path to your local MCP server build.
+> Make sure the `KONNECT_ACCESS_TOKEN` matches the one you set earlier.
```
Comment (cloudjumpercat) [inline app/_how-tos/get-started-with-mcp.md:175]:
```suggestion
> * Make sure the `KONNECT_ACCESS_TOKEN` matches the one you set earlier.
> * Replace the `KONNECT_REGION` value with your geographic region. You can see a list of all {{site.konnect_short_name}} regions in the [Geographic regions](/konnect-platform/geos/#control-planes) documentation.
```

File: app/_how-tos/get-started-with-mcp.md [hunk 21]
```diff
@@ -0,0 +1,372 @@
...
+        "KONNECT_ACCESS_TOKEN": "kpat_api_key_here",
+        "KONNECT_REGION": "us"
+      }
+    }
+  }
+}
+```
+
+{:.warning}
+> Replace `/absolute/path/to/mcp-konnect/build/index.js` with the full path to your local MCP server build.
```
Comment (cloudjumpercat) [inline app/_how-tos/get-started-with-mcp.md:174]:
```suggestion
> * Replace `/absolute/path/to/mcp-konnect/build/index.js` with the full path to your local MCP server build.
```

File: app/_how-tos/get-started-with-mcp.md [hunk 22]
```diff
@@ -0,0 +1,372 @@
...
+
+tools:
+    - deck
+
+prereqs:
+  entities:
+    services:
+        - example-service
+    consumers:
+        - example-consumer
```
[the remark this answers is not in the document]
Comment (tomek-labuk) [reply app/_how-tos/get-started-with-mcp.md:50]: Fixed

File: app/_how-tos/get-started-with-mcp.md [hunk 23]
```diff
@@ -0,0 +1,354 @@
...
+
+### List all Control Planes
+
+{% navtabs "list-all-control-planes" %}
+{% navtab "Prompt" %}
+
+Use this prompt to retrieve all control planes in your {{site.konnect_short_name}} organization:
+
+```text
+List all control planes in my Kong {{site.konnect_product_name}} organization.
```
Comment (cloudjumpercat) [inline app/_how-tos/get-started-with-mcp.md:176]:
```suggestion
List all Control Planes in my {{site.konnect_product_name}} organization.
```

File: app/_how-tos/get-started-with-mcp.md [hunk 24]
```diff
@@ -0,0 +1,354 @@
...
+      icon_url: /assets/icons/gateway.svg
+    - title: Destroy the {{site.base_gateway}} container
+      include_content: cleanup/products/gateway
+      icon_url: /assets/icons/gateway.svg
+
+min_version:
+    gateway: '3.6'
+
+faqs:
+  - q: Why am I getting a connection error?
```
Comment (cloudjumpercat) [inline app/_how-tos/get-started-with-mcp.md:82]:
```suggestion
  - q: "Why am I getting a connection error for the {{site.konnect_short_name}} MCP server?"
```
Just to increase findability in searches.

File: app/_how-tos/get-started-with-mcp.md [hunk 25]
```diff
@@ -0,0 +1,354 @@
...
+    a: |
+      * **Verify your API key** is valid and has the necessary permissions.
+      * **Check the API region** is correctly specified (`KONNECT_REGION`).
+      * **Ensure network access** to the Kong {{site.konnect_product_name}} API is not blocked by a firewall or proxy.
+
+  - q: Why am I seeing authentication errors?
```
Comment (cloudjumpercat) [inline app/_how-tos/get-started-with-mcp.md:88]:
```suggestion
  - q: "Why am I seeing authentication errors for the {{site.konnect_short_name}} MCP server?"
```

File: app/_how-tos/get-started-with-mcp.md [hunk 26]
```diff
@@ -0,0 +1,354 @@
...
+    a: |
+      * **Regenerate your API key** from the Kong {{site.konnect_product_name}} portal if it's expired or revoked.
+      * **Confirm environment variables** like `KONNECT_ACCESS_TOKEN` are set and available to the process.
+
+  - q: Why is my data not found?
```
Comment (cloudjumpercat) [inline app/_how-tos/get-started-with-mcp.md:93]:
```suggestion
  - q: "Why is my data not found when using the {{site.konnect_short_name}} MCP server?"
```

**Review by cloudjumpercat — CHANGES_REQUESTED** [review]
Left a few suggestions. For some reason, all of the "bring your own MCP server" links seem to be broken (404s). I can't figure out why since the permalinks match what I'm searching for.

File: app/_how-tos/govern-mcp-traffic.md [hunk 27]
```diff
@@ -0,0 +1,457 @@
...
+message: The repository YOUR_REPOSITORY_NAME has {n} active branches.
+{% endvalidation %}
+<!-- vale on -->
+
+{% endnavtab %}
+{% endnavtabs %}
+
+### Denied requests
+
+Each input below matches a deny pattern like `.*(backdoor|exfiltrate|CVE-\d{4}-\d+).*`, which should trigger rejection by the AI Prompt Guard plugin. Let me know if you'd like boundary cases or false positives to test.
```
[the remark this answers is not in the document]
Comment (cloudjumpercat) [reply app/_how-tos/govern-mcp-traffic.md:276]: I think this sentence still needs to be removed.

File: app/_how-tos/govern-mcp-traffic.md [hunk 28]
```diff
@@ -0,0 +1,446 @@
...
+Request #1 — Status: 200
+Request #2 — Status: 429
+Request #3 — Status: 429
+Request #4 — Status: 200
+Request #5 — Status: 429
+Request #6 — Status: 429
+```
+{:. no-copy-code  }
+
+Our rate limit configuration is allowing 1 requests per 60 seconds, which means that:
```
Comment (cloudjumpercat) [inline app/_how-tos/govern-mcp-traffic.md:434]:
```suggestion
Our rate limit configuration is allowing one request per 60 seconds, which means that:
```

File: app/_how-tos/secure-mcp-traffic.md [hunk 29]
```diff
@@ -0,0 +1,213 @@
...
+title: "Secure GitHub MCP Server traffic with {{ site.base_gateway }} and AI Gateway"
+content_type: how_to
+related_resources:
+  - text: AI Gateway
+    url: /ai-gateway/
+  - text: AI Proxy Advanced
+    url: /plugins/ai-proxy-advance/
+  - text: Key Auth plugin
+    url: /plugins/key-auth/
+permalink: /mcp/secure-mcp-traffic
```
Comment (cloudjumpercat) [inline app/_how-tos/secure-mcp-traffic.md:11]:
```suggestion
permalink: /mcp/secure-mcp-traffic/
```

File: app/_how-tos/observe-mcp-traffic.md [hunk 30]
```diff
@@ -0,0 +1,204 @@
...
+related_resources:
+  - text: AI Gateway
+    url: /ai-gateway/
+  - text: AI Proxy Advanced
+    url: /plugins/ai-proxy-advanced/
+  - text: Prometheus plugin
+    url: /plugins/prometheus/
+  - text: Monitor AI LLM metrics
+    url: /ai-gateway/monitor-ai-llm-metrics/
+permalink: /mcp/observe-mcp-traffic
```
Comment (cloudjumpercat) [inline app/_how-tos/observe-mcp-traffic.md:13]:
```suggestion
permalink: /mcp/observe-mcp-traffic/
```

**Review by cloudjumpercat — APPROVED** [review]
Approved, just needs the rebase cleaned up

**Comment by tomek-labuk** [comment]
Resolved in: https://github.com/Kong/developer.konghq.com/pull/2006
